cli/gen-package/src/Generate.elm

```diff
@@ -555,26 +560,37 @@ asArgumentTypeHelper tipe =
 {-| -}
 asValue : Int -> Elm.Type.Type -> Elm.Expression
 asValue index tipe =
-    getArgumentUnpacker tipe <|
+    getArgumentUnpacker 0 tipe <|
         Elm.valueFrom
             local
             (argName index)
 
 
-getArgumentUnpacker : Elm.Type.Type -> Elm.Expression -> Elm.Expression
-getArgumentUnpacker tipe value =
+getArgumentUnpacker : Int -> Elm.Type.Type -> Elm.Expression -> Elm.Expression
+getArgumentUnpacker freshCount tipe value =
     case tipe of
-        Elm.Type.Lambda _ two ->
+        Elm.Type.Lambda one two ->
             let
                 f =
-                    getArgumentUnpacker two value
+                    getArgumentUnpacker (freshCount + 1)
+                        two
+                        (Elm.apply value [ Elm.value varName ])
+
+                varName =
+                    "lambdaArg" ++ String.fromInt freshCount
             in
-            Elm.apply f [ ElmGen.pass ]
+            ElmGen.lambdaBetaReduced (Elm.string varName)
+                (typeToExpression one)
+                (\_ ->
+                    Elm.lambdaBetaReduced varName
+                        (typeToAnnotation one)
+                        (\_ -> f)
+                )
```

Comment:
The critical insight (does it merit a comment?) is that you:
1) lambda (beta reduced) with `varName` both in the generated binding and the code that the binding generates
2) lambda with `varName` "outside" (after calling the unpacker for the inner type) but apply it "inside" (before calling the unpacker)